File: accudrop/app/src/main/java/me/chrislane/accudrop/presenter/ReplayMapPresenter.java
package me.chrislane.accudrop.presenter;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.location.Location;
import android.support.v4.app.Fragment;
import android.support.v4.util.Pair;

import java.util.List;
import java.util.UUID;

import me.chrislane.accudrop.fragment.ReplayMapFragment;
import me.chrislane.accudrop.viewmodel.ReplayViewModel;

public class ReplayMapPresenter {

    private final ReplayMapFragment replayMapFragment;
    private final Fragment parentFragment;
    private ReplayViewModel replayViewModel;

    public ReplayMapPresenter(ReplayMapFragment replayMapFragment) {
        this.replayMapFragment = replayMapFragment;

        parentFragment = replayMapFragment.getParentFragment();
        if (parentFragment != null) {
            replayViewModel = ViewModelProviders.of(parentFragment).get(ReplayViewModel.class);
        }
    }

    /**
     * Add all observers.
     */
    public void addObservers() {
        subscribeToRoute();
    }

    /**
     * Set the new route being displayed.
     */
    private void subscribeToRoute() {
        final Observer<List<Pair<UUID, List<Location>>>> routeObserver = replayMapFragment::updateMapRoutes;
        replayViewModel.getUsersAndLocs().observe(parentFragment, routeObserver);
    }
}

File: app/src/main/java/me/chrislane/accudrop/presenter/ReplayMapPresenter.kt
package me.chrislane.accudrop.presenter

import android.location.Location
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import me.chrislane.accudrop.fragment.ReplayMapFragment
import me.chrislane.accudrop.viewmodel.ReplayViewModel
import java.util.*

class ReplayMapPresenter(private val replayMapFragment: ReplayMapFragment) {
    private val parentFragment: Fragment? = replayMapFragment.parentFragment
    private var replayViewModel: ReplayViewModel

    init {

        if (parentFragment == null) {
            throw Exception("Missing parent fragment (Replay)")
        }
        replayViewModel = ViewModelProvider(parentFragment).get(ReplayViewModel::class.java)
    }

    /**
     * Add all observers.
     */
    fun addObservers() {
        subscribeToRoute()
    }

    /**
     * Set the new route being displayed.
     */
    private fun subscribeToRoute() {
        val routeObserver = Observer<MutableList<Pair<UUID, MutableList<Location>>>> {
            if (it != null) {
                replayMapFragment.updateMapRoutes(it)
            }
        }
        replayViewModel.getUsersAndLocs().observe(parentFragment!!, routeObserver)
    }
}
